unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp

```diff
@@ -0,0 +1,274 @@
+          "java::generics$bound_element::NUM");
+        REQUIRE(
+          java_generics_class_type_var(0, java_generics_class_type)==
+          "java::generics$bound_element::NUM");
+
+        typet &sub_type=generic_type_var.subtype();
+        REQUIRE(sub_type.id()==ID_symbol);
+        symbol_typet &bound_type=to_symbol_type(sub_type);
+        REQUIRE(bound_type.get_identifier()=="java::java.lang.Number");
+        REQUIRE(
+          to_symbol_type(java_generics_class_type_bound(0, java_generics_class_type)).get_identifier()==
+          "java::java.lang.Number");
+
+        const struct_union_typet::componentt &elem=
+          java_generics_class_type.get_component("elem");
+        const typet &elem_type=java_class_type.component_type("elem");
+
+        REQUIRE(is_java_generic_parameter(elem_type));
+      }
+    }
+  }
+
+  GIVEN("Some class files with generic type variable")
+  {
+    WHEN("Parsing a class with bounded type variable")
+    {
+      java_lang->parse(java_code_stream, "generics.class");
+
+      symbol_tablet new_symbol_table;
+      java_lang->typecheck(new_symbol_table, "");
+
+      java_lang->final(new_symbol_table);
+
+      REQUIRE(new_symbol_table.has_symbol("java::generics"));
+      REQUIRE(new_symbol_table.has_symbol("java::generics$element"));
+      REQUIRE(new_symbol_table.has_symbol("java::generics$bound_element"));
+
+      THEN("The generic fields should be annotated with concrete types")
+      {
+        const symbolt &class_symbol=new_symbol_table.lookup("java::generics");
+        const typet &symbol_type=class_symbol.type;
+
+        REQUIRE(symbol_type.id()==ID_struct);
+        class_typet class_type=to_class_type(symbol_type);
+        REQUIRE(class_type.is_class());
+        java_class_typet java_class_type=to_java_class_type(class_type);
+        REQUIRE(!is_java_generics_class_type(java_class_type));
+
+        const struct_union_typet::componentt &belem=java_class_type.get_component("belem");
+        const typet &belem_type=java_class_type.component_type("belem");
+
+        REQUIRE(belem_type!=nil_typet());
+        REQUIRE(is_java_generic_type(belem_type));
+
+        const java_generic_typet &container=
+          to_java_generic_type(belem_type);
+
+        const std::vector<java_generic_parametert> &generic_types=container.generic_type_variables();
+        REQUIRE(generic_types.size()==1);
+
+        const typet& inst_type=java_inst_get_inst_type(0, container);
+
+        REQUIRE(inst_type.id()==ID_pointer);
+        const typet &inst_type_symbol=inst_type.subtype();
+        REQUIRE(inst_type_symbol.id()==ID_symbol);
+        REQUIRE(
+          to_symbol_type(inst_type_symbol).get_identifier()==
+          "java::java.lang.Integer");
+      }
+    }
+  }
+
+  GIVEN("Some class files with Generics")
+  {
+    WHEN("Parsing a class with generic return type of methods")
+    {
+      java_lang->parse(java_code_stream, "generics$bound_element.class");
```

Comment:
Overlap with the other test that parses this file to keep tests together and to reduce duplication